### PR TITLE
Namespace as kind

### PR DESCRIPTION
**Pull request by derekwaynecarr**
A `Namespace` object is defined to support discovery and traversal of Kubernetes cluster.

**Model changes:**

A `Namespace` contains the following attributes:
1. Name
2. Labels
3. Annotations

A `Namespace` is not namespace scoped.

**kube-apiserver**
- Kubernetes creates a `default` namespace by default and ensures it exists.

**REST API**

A `Namespace` operates like any other resource in Kubernetes.

Like `Node`, it has a cluster-scope.

HTTP Semantics:

| Action | HTTP Verb | Path | Description |
| --- | --- | --- | --- |
| CREATE | POST | /api/{version}/namespaces/ | Create namespace |
| GET | GET | /api/{version}/namespaces/{name} | Get namespace {name} |
| UPDATE | PUT | /api/{version}/namespaces/{name} | Update namespace {name} |
| DELETE | DELETE | /api/{version}/namespaces/{name} | Delete namespace {name} |
| LIST | GET | /api/{version}/namespaces/ | List all namespaces |
| WATCH | GET | /api/{version}/watch/namespaces/ | Watch all namespaces |

This PR modifies the v1beta3 URL structure for resources that are scoped to a namespace.

| URL | Description |
| --- | --- |
| /api/{version}/namespaces/{namespace}/{resource}/{name} | Fetch the resource of type {resource} with name {name} in namespace {namespace} |

**Data Storage**

A `Namespace` object is not in a namespace, and is stored in a etcd key-space `/registry/namespace/{name}` to ensure uniqueness.

**Client**
`kubectl` can list namespaces

``` shell
$ kubectl get namespaces
```

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
Resolves #2159

**Comment by derekwaynecarr** [comment]
Items remaining:

1. Test cases
2. Ability to enforce existence of a namespace prior to creating resources in that namespace
3. Ability to clean-up resources when namespace is marked for deletion.

**Comment by derekwaynecarr** [comment]
Added 2 admission control plug-ins to control behavior as deployment choice:

1. NamespaceExists - requires predeclaration of a namespace before usage
2. NamespaceAutoProvision - automatically creates the namespace upon first usage

NamespaceAutoProvision is probably best used pre-1.0 for upstream to maintain backwards compatibility for current behavior.

**Comment by derekwaynecarr** [comment]
Just need to add a few more unit tests.  Will handle namespace clean-up process in a follow-on PR, but want to keep this initial PR to a minimum set.

**Comment by erictune** [comment]
I think I like the semantics you describe.  It looks like it allows creating sort of hierarchies. 

I don't see the point of making the "NamespaceExists" check optional.  If we think this is the right thing to do, shouldn't we do it for everyone?  Otherwise we have to reason about multiple different semantics for something which is central to the system.  And it won't get any easier to make it mandatory than it is today.

**Comment by derekwaynecarr** [comment]
I included the auto-provision use case for backwards compatibility, and for you guys, where I thought you were not always intent on requiring pre-declaration prior to use.  I thought @jbeda noted this explicitly at our recent meet-up.

Basically, I didn't want this PR to get held up this week because it would be a breaking change.  We have downstream use cases for this in the next couple weeks, that I wanted to just get the basics in this week if possible.

If at 1.0, or some future time, we mandate NamespaceExists, then I wouldn't even make it a command line argument choice and instead just have it be an implicit admission handler.

**Comment by thockin** [comment]
In your semantics, as per the table above, what is context?
`/api/{version}/ns/{context}/namespaces	`

I don't understand the `/api/{version}/ns/{ns}/namespaces/{ns}` construct.  Why have {ns} twice, as opposed to just `/api/{version}/ns/{ns}`?  Can the different {ns} placeholders be different names?  It feels like this is a workaround to a problem I don't see?

**Comment by derekwaynecarr** [comment]
I ended up viewing the namespace object a lot like the minion object.

Here is the bootstrapping scenario I was working towards:

Alice admin does a ```cluster/kube-up.sh``` to create a new cluster.

The cluster has a ```default``` namespace with our default pod, service, replication controller, policy [future] items.

Alice wants to see what namespaces exist.  If the policy in the ```default``` namespace had an ALLOW rule to LIST namespaces, the following would be accepted.

```
cluster/kubectl.sh get namespaces  
Request: GET /api/{version}/ns/default/namespaces
Response: []Namespace{ {Name: "default"} }
```

Alice wants to create a new namespace ```foo```.  If the policy in the ```default``` namespace had an ALLOW rule to create new namespaces, the following would be accepted.

```
cluster/kubectl.sh create -f foo_namespace.json 
Request: POST /api/{version}/ns/default/namespaces
```

Alice verifies her namespace was created by listing them from the default context.

```
cluster/kubectl.sh get namespaces  
Request: GET /api/{version}/ns/default/namespaces
Response: []Namespace{ {Name: "default"}, {Name: "foo"} }
```

Alice switches to now work in the ```foo``` namespace context.

```
cluster/kubectl.sh namespace foo
Using namespace foo
```

**FUTURE: This is when policy is persisted**
Alice wants to restrict users ability to create new namespaces from ```foo```.
Alice wants to restrict users ability to list other namespaces from ```foo```.
Alice wants to give Bob limited access to ```foo``` to create pods, etc.

```
cluster/kubectl.sh create -f policy.json
Request: POST /api/{version}/ns/foo/policies
Response: 201
```

Alice tries to list namespaces from ```foo```, but the policy rejects it:

```
cluster/kubectl.sh get namespaces  
Request: GET /api/{version}/ns/foo/namespaces
Response: 403 Forbidden
```

Alice tries to create a new namespace from ```foo``` namespace context:

```
cluster/kubectl.sh create -f foo_namespace.json 
Request: POST /api/{version}/ns/foo/namespaces
Response: 403 Forbidden
```

So in this case ```context``` in my original discussion was the requesting namespace scope.  The pattern of ```/ns/context/namespaces``` was proposed to provide a policy evaluation context.

Coming at this from the experience of bootstrapping a cluster and locking down who could or could not create and list namespaces from the initial bootstrapped ```default``` namespace is how I arrived at the current proposal.

Note internally, if you want to LIST/WATCH all namespaces to support controllers, we would support the following:

```
GET /api/{version}/namespaces  (LIST all namespaces)
GET /api/{version}/watch/namespaces (WATCH all namespaces)
```

I would plan to use the above APIs to support a controller to handle clean-up of resources in a namespace when a namespace is marked for deletion.

Hope this helps understand what is proposed.

**Comment by smarterclayton** [comment]
I also had the confusion.  Is namespace a security context?  I hadn't viewed it as such before - instead I had assumed that the policy engine made a decision on whether you could create namespaces based on other criteria (since you don't create namespaces into another namespace).  I don't know that I follow the subtlety of the reasoning, so let me look at it some more.

**Comment by thockin** [comment]
Thanks for the explanation.  I now understand what you did.  It is a bit
tricky (and a bit elegant).  In effect, creating a namespace "under" a
context really creates a global namespace IFF the context is allowed to do
so.  So if I create /api/.../ns/foo/namespaces/bar, it also magically
manifests as /api/.../ns/default/namespaces/bar ?

How is this reflected in etcd storage?

**Comment by derekwaynecarr** [comment]
@thockin - in etcd storage, namespaces are stored as ```/registry/namespace/<name>```.  This is why I say a ```Namespace``` does not have a ```namespace```.  It is a globally unique name to the etcd repository.  I do not intend to nest ```Namespace``` objects.  I think that leads to insanity in short term.

If you guys are not comfortable with using the security context of one namespace to allow you to derive or list others, I am fine with modifying this PR to instead just have a path structure like:

```
POST /api/version/ns
BODY <namespace object>
Result: create a namespace

GET /api/version/ns
Result: list all namespaces

GET /api/version/ns/{ns}
Result: get namespace named {ns}
```

Note if we go this path, the HTTP semantics are nice, but the policy around who can create these things will need to be defined in a Policy related PR.  I just went this route since I asked what would I expect to happen from the bootstrapping scenario.  

The alternate approach will require some special logic to map a ```RESTStorage``` to these base paths, but I think that is not a major concern and can be worked out.  How ```kubectl``` handles this resource type will have some special behavior, but I also do not think that is a major issue and should not be a problem.

I am going to let the PR stay as currently defined, but can we try to reach consensus on what we want to do with the approach from @smarterclayton, @thockin, @erictune by end of week after you have had some time to reflect?  

I would like to get this basic support in Kubernetes early next week to make some progress on issues specific to OpenShift Origin that this blocks.  In interim, I can look at some other work.

Given that @erictune has thought a lot on ```Policy```, I would be interested in getting his preference, and trust his recommendation assuming he understood what I was initially proposing.

**Comment by derekwaynecarr** [comment]
>>In effect, creating a namespace "under" a context really creates a global namespace IFF the context is allowed to do so.  So if I create /api/.../ns/foo/namespaces/bar, it also magically manifests as /api/.../ns/default/namespaces/bar ?

Yes - this is why it's equivalent to ```Node```.

If I create a ```Node```, it is addressable from any namespace context because it is global.

As a result, both of these requests return the same value today.

```
GET /api/version/ns/default/nodes/node1
GET /api/version/ns/foo/nodes/node1
```

I imagined the ability to CRUD Nodes (or any global resource like a Namespace), would follow the policy pattern outlined.

**Comment by smarterclayton** [comment]
I'm provisionally ok with it, especially since bootstrapping overlap with a policy engine has some tricky bits (which I agree with in your outline).

Imagine a scenario where you have a very large cluster.  You want to subdivide parts of the cluster such that user level cluster tools can interact with the cluster and "enhance" the experience for some subsets (ie you have a PaaS running across certain nodes and namespaces).  The infrastructure for that tool is running on the root infrastructure as pods in a namespace (call it "dev-infra"), and has certain elevated privileges on a subset of Kube nodes, namespaces, and resources.  The pods need to be able to act against the master in some context that is not "root", but is nevertheless elevated for that subset of resources.  With a sufficiently rich policy engine you can define rules that apply to actors in "dev-infra" that control "dev1", "dev2", "dev3", etc.  The policy for the admins of dev-infra and the pods themselves looks very similar - perhaps defined on that namespace, but reflected in the policy engine running in "default" with more sophisticated delegation rules.  The admin that creates "dev1" is doing so from within "dev-infra", so "dev1" is given a descendant policy from "dev-infra".  The policy that allowed that action is the parent policy of "dev1", etc.

Namespace is a nice "context" simplification - it only goes so far, but it covers a lot of space.

Also, namespaces might need to be shardable, so having their mutation be a sub path of a parent path is superior to making actions on the parent path special.  It also normalizes tools - no special logic for create/update/delete.

**Comment by thockin** [comment]
Yeah, I am pretty OK with it (though I think we need a doc with a really
great explanation that I can refer back to when I forget how it works).
I'll defer to Eric, though, since he holds context on policy stuff.

**Comment by erictune** [comment]
@derekwaynecarr 

TL;DR: please switch to the form you suggested in your next-to-last comment.  Expressing the policy for bootstrapping and the use cases your mention won't be a problem.  I'll gladly merge after that.

Long version:

There are some nice properties with the `/api/{version}/ns/{foo}/namespaces/{bar}` form, but interesting could also be confusing for users, as evidenced by the length of the conversation on this thread.

I think that the namespaced nodes thing we have is sort of funky.  I think I'd like it better if, at some point, we deprecated the current `/api/version/ns/ANYTHING/nodes/node1` format and just made it like `/api/version/nodes/node1`.  (Different PR, of course, and perhaps different author).

Likewise, that would suggest a form like this for namespace operations: `/api/version/ns/{ns}`.  You offered to do that above, and I prefer that you do. 

So, then that leaves the question of how to handle bootstrapping and the other use cases you mentioned.  Here are my suggestions:
 - At cluster creation, Alice admin has a policy like `{user: "alice"}`.  This matches everything.  So Alice can do anything, including create or list a namespace.    That should work now IIUC.
 - Later, we may wish to teach the `KindAndNamespace` function in `pkg/apiserver/handlers.go` that an url like   `/api/version/ns/foo` has attributes `{Kind: "ns", Namespace: "", Name: "foo"}` and that `/api/version/node/bar` has attributes `{Kind: "node", Namespace: "", Name: "bar"}`.  Where empty string and undefined are the same for policy evaluation.
 - You mentioned the use case: *Alice wants to restrict users ability to create new namespaces from foo.*.  I'd rephrase that use case as *Alice wants to restrict foo-team's ability to create new namespaces*, which would be implemented as this policy `{Kind: "ns", Group: "foo-team", ReadOnly: true}`
 - You mentioned the use case: *Alice wants to restrict users ability to list other namespaces from foo*.  I think by that you really mean these three use cases:
  - *Alice wants to restrict foo-team's ability to list all namespaces*.  That would require no particular policy.
  - *Alice wants to allow foo-team to get the properties of namespace "foo"*.  That would have this policy: `{Kind: "ns", Group: "foo-team", ReadOnly: true, Name: "foo"}`.
  -  *Alice wants to allow foo-team to list the several namespaces they have access to, e.g. "foo1", "foo2", "foo3".   I don't have a solution for that yet.  I don't think this PR does either.   Let's defer that question.  
 - You mentioned the use case: *Alice wants to give Bob limited access to foo to create pods, etc*.  We already know we can do that.

**Comment by smarterclayton** [comment]
The problem with this is every client needs to know whether opaque resource types are namespaced or not.  Right now, the client can assume that /ns/{} can be specified without the server rejecting it (i.e., the server can handle that context problem).  If you omit /ns/{} from nodes, then the client has to know whether a resource is namespaced before it tries to POST it.  If only namespace follows this behavior, that's one thing.  But if nodes and other namespaceless resources start being used, every client has to know the difference.  Also, in the future, what if we *do* decide to namespace nodes?  Then old clients are broken.

I don't think of namespace as "aspect of resource" - I think of it as "fundamental server context".  Having to know whether "fundamental server context" is required for a call or not complicates clients.

----- Original Message -----

**Comment by derekwaynecarr** [comment]
I think the discussion of tools, clients, etc should move away from this PR and go on #3806.

I noted I was ok with either approach last week, so I do not want to go back on that statement.  I have rewritten a lot of code in this project, so I am fine to continue doing that ;-)

>>*Alice wants to allow foo-team to list the several namespaces they have access to, e.g. "foo1", "foo2", "foo3". I don't have a solution for that yet. I don't think this PR does either. Let's defer that question.

I am curious if you think this is a question we will attempt to answer soon, pre or post v1, and if the answer to that question should be part of #3806.

**Comment by smarterclayton** [comment]
Derek pointed out very reasonably that admin resources do not necessarily have to be namespace scoped (they could be bound to a namespace) in which case they are likely outside the common client path and called from a different cli than kubectl (which is aggressively end-user focused).  So I'm reasonably ok with assuming everything kubectl accesses is namespace scoped, and that any server level resource that should be exposed to end users should also be namespaced scoped (i.e. if you want to expose nodes so that a user can see it, it's different than nodes that an admin sees).

----- Original Message -----

**Comment by erictune** [comment]
Derek -- some thoughts regarding this use case: *Alice wants to allow foo-team to list the several namespaces they have access to, e.g. "foo1", "foo2", "foo3".*

In hosted clusters and mature private clusters, namespace creation will often be tied to a workflow that includes setting up billing and/or other bureaucratic stuff.  So, there will probably already be some kind of proxy in the form of a web service where you acquire and view namespaces.   
For example, the `gcloud` tool for gce does not have a command to list projects.  You have to go to a human-only web page to see your GCE projects.  So, it seems reasonable to require foo-team to use some proxy to find out what namespaces it can access. 

Another way to answer the question of what namespaces I have "access" to is to list all the policy objects that mention me as a principal and then grep and uniq all the namespaces you saw.  This is related to the ResourceAccessReview that @deads2k is working on. 

At any rate, I don't see this as urgent to figure out, because anyone paranoid about exposing namespace names is willing to do some work to implement a proxy as described above, and everyone else can just have permissive policy.

**Comment by derekwaynecarr** [comment]
Thanks @erictune - I am trying to coordinate what we will do in OpenShift for that use case in the next few weeks.  An OpenShift specific proxy is fine.

**Comment by derekwaynecarr** [comment]
I have updated this PR to support the URL patterns requested by @erictune 

As part of this PR, I had to modify ```RESTStorage``` and ```kube-apiserver```  to handle cluster scoped resources.  I verified that Swagger UI shows the proper resource paths.

I will look to add ```kubectl``` support for listing namespaces in a follow-on PR because I think the UI in that piece will be more interesting.

At this point, you can curl just fine.

**Comment by derekwaynecarr** [comment]
@erictune - I think this is ready for review.  
@smarterclayton - your feedback re: the ```RESTStorage``` changes are needed too.

**Comment by derekwaynecarr** [comment]
There are some unique things about this PR that I want to investigate further.

Specifically:

1. What is the impact of having a resource path (/ns) that is not a plural form of meta.Kind (Namespace)?
2. Think through some on how kubectl will deal with the ramification of 1
3. Look to see if Objects that are external to a Namespace should embed a different struct so RESTMapper can know to include namespace params in client requests.

**Comment by derekwaynecarr** [comment]
I think there are some prereqs that need to be fulfilled before doing this work:

1. Move ```ObjectMeta.Namespace``` to ```NamespaceMeta.Namespace```
2. Types that have a ```Namespace``` should embed ```NamespaceMeta```
3. Types that do not have a ```Namespace``` should not embed ```NamespaceMeta```
4. Update conversion routines for 2 and 3.
5. Modify ```meta.Interface``` to have a ```RequireNamespace() bool``` that returns ```true``` if the type embedded ```NamespaceMeta``` else ```false```.
6. Update ```kubectl``` to not set a ```NamespaceParam``` if the resource does not ```RequireNamespace```

Once the above are done, we need to revisit this discussion:

Do we want a Kind called ```Namespace``` or ```Ns```?

This has ramifications on the URL paths and how we map resource to the plural form of Kind.

| Kind | Resource | Sample URLS |
| ------ | ------------- | -------------------- |
| Namespace | namespaces | /namespaces/{namespace}/pods/{pod} |
| NS | ns | /ns/{ns}/pods/{pod} |

This has a ramification from the CLI:

kubectl get namepsaces foo

or

kubectl get ns foo

I know downstream OpenShift needs the ability to better distinguish resources that are outside of a namespace scope per discussion with @liggitt

Thoughts on this from others?

**Comment by derekwaynecarr** [comment]
A slight variation on the above is the following:

1. Move ```ObjectMeta.Namespace``` into ```NSObjectMeta```
2. ```NSObjectMeta``` would inline ```ObjectMeta```
3. Types that have a namespace would embed ```NSObjectMeta```, types that do not would embed ```ObjectMeta```.

In this way, our JSON api can continue to map ```ObjectMeta``` or ```NSObjectMeta``` to ```{ "metadata": {...}``` for either type.

**Comment by erictune** [comment]
I was imagining a Kind called Namespace, which does solve the plural issue.  No strong preference though.

**Comment by erictune** [comment]
I thought about work items 1, 2, 3 and 4 before I suggested you make namespaces not-namespaced.
I figured that we would continue to use ObjectMeta but by convention leave it empty for non-namespaced objects.  For item 5, I imagined that we would just have a hardcoded list of kinds in the RequireNamespace() function, and a separate implementation of that in each `pkg/api/$VERSION/`, which would allow us to change the RequiresNamespace of a Kind with a version change.

**Comment by erictune** [comment]
Oh, I see you already started on 1 and 2.  Okay.

**Comment by derekwaynecarr** [comment]
@erictune - after speaking with @smarterclayton , I am puting 1 and 2 on hold, and will document an alternate path that is less disruptive today, but still should make everything work nicely.

**Comment by derekwaynecarr** [comment]
This is ready for final review.

I updated the original PR description to document what is done.

**tl,dr:**

1. namespace is cluster-scoped
2. v1beta3 url formats are /api/version/namespaces/namespace/pods/name
3. kubectl works fine

Always enabling the plug-in to enforce existence of a namespace prior to usage can be done in a follow-on PR, but vagrant cluster requires the namespace to exist now.

**Comment by derekwaynecarr** [comment]
Rebased

**Comment by erictune** [comment]
Will look at monday when I return to work.  Or feel free to reassign to
Clayton for faster merge.

**Comment by abonas** [comment]
this is a bit confusing (from one of the examples above): "GET /api/{version}/ns/default/namespaces" - getting all namespaces from a namespace?

**Comment by derekwaynecarr** [comment]
That was an earlier iteration of the work.  The api is as presented in the main PR description.

Get /api/version/namespaces will just return all namespaces as expected.

**Comment by smarterclayton** [comment]
No other comments besides those above.

**Comment by derekwaynecarr** [comment]
Code review comments completed.

**Comment by derekwaynecarr** [comment]
Rebased again.

File: pkg/api/rest/types.go [hunk 1]
```diff
@@ -134,3 +134,30 @@ func (nodeStrategy) Validate(obj runtime.Object) errors.ValidationErrorList {
 	node := obj.(*api.Node)
 	return validation.ValidateMinion(node)
 }
+
+// namespaceStrategy implements behavior for nodes
```
Comment (erictune) [inline pkg/api/rest/types.go]: ...for namespaces.

File: pkg/api/v1beta1/conversion.go [hunk 2]
```diff
@@ -731,6 +731,21 @@ func init() {
 			}
 			return nil
 		},
+		func(in *Namespace, out *newer.Namespace, s conversion.Scope) error {
```
Comment (erictune) [inline pkg/api/v1beta1/conversion.go]: This is different from other funcs above.  Is it correct, and if so does it deserve a comment?

File: pkg/client/namespaces_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+*/
+
+package client
+
+import (
+	"testing"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	//"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
```
Comment (erictune) [inline pkg/client/namespaces_test.go]: remove this line

**Comment by erictune** [comment]
very minor comments, and needs squash.  then LGTM.

File: pkg/api/v1beta1/conversion.go (hunk 2 above)
Comment (derekwaynecarr) [inline pkg/api/v1beta1/conversion.go]: I am not sure what is different in your view, but I believe it's correct

**Comment by derekwaynecarr** [comment]
squashed commits, addressed last round of comments.

**Comment by smarterclayton** [comment]
LGTM

**Comment by smarterclayton** [comment]
Rerunning travis before merging
